kopf/clients/auth.py

```diff
@@ -1,42 +1,125 @@
+
+    if verify:
+        verify_pykube()
+
+
+def login_client(verify=False):
+    import kubernetes.client
     try:
         kubernetes.config.load_incluster_config()  # cluster env vars
-        logger.debug("configured in cluster with service account")
+        logger.debug("Client is configured in cluster with service account.")
     except kubernetes.config.ConfigException as e1:
         try:
             kubernetes.config.load_kube_config()  # developer's config files
-            logger.debug("configured via kubeconfig file")
+            logger.debug("Client is configured via kubeconfig file.")
         except kubernetes.config.ConfigException as e2:
-            raise LoginError(f"Cannot authenticate neither in-cluster, nor via kubeconfig.")
+            raise LoginError(f"Cannot authenticate client neither in-cluster, nor via kubeconfig.")
 
-    # Make a sample API call to ensure the login is successful,
-    # and convert some of the known exceptions to the CLI hints.
+    if verify:
+        verify_client()
+
+
+def verify_pykube():
+    """
+    Verify if login has succeeded, and the access configuration is still valid.
+    """
+    try:
+        api = get_pykube_api()
+        rsp = api.get(version="", base="/")
+        rsp.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        if e.response.status_code == 401:
+            raise AccessError("Cannot authenticate to the Kubernetes API. "
+                              "Please login or configure the tokens.")
+        else:
+            raise
+
+
+def verify_client():
+    """
+    Verify if login has succeeded, and the access configuration is still valid.
+    """
+    import kubernetes.client.rest
     try:
         api = kubernetes.client.CoreApi()
         api.get_api_versions()
     except urllib3.exceptions.HTTPError as e:
-        raise LoginError("Cannot connect to the Kubernetes API. "
-                         "Please configure the cluster access.")
+        raise AccessError("Cannot connect to the Kubernetes API. "
+                          "Please configure the cluster access.")
     except kubernetes.client.rest.ApiException as e:
         if e.status == 401:
-            raise LoginError("Cannot authenticate to the Kubernetes API. "
-                             "Please login or configure the tokens.")
+            raise AccessError("Cannot authenticate to the Kubernetes API. "
+                              "Please login or configure the tokens.")
         else:
             raise
+
+
+def get_pykube_cfg() -> pykube.KubeConfig:
+    if _pykube_cfg is None:
+        raise LoginError("Not logged in with PyKube.")
+    return _pykube_cfg
+
+
+# TODO: add some caching, but keep kwargs in mind. Maybe add a key= for purpose/use-place?
+def get_pykube_api(timeout=None) -> pykube.HTTPClient:
+    return pykube.HTTPClient(get_pykube_cfg(), timeout=timeout)
```

Comment:
What about specifying a non-infinite timeout by default? Pykube's default is 10s, but I think explicitly passing `None` here would deactivate that reasonable default.

Reply:
Fixed.

Reply:
This behaviour caused some issues with 0.19rc1 and will be reverted to no-timeouts for all internal api calls, and especially the watch-requests — same as it was before migration to pykube: https://github.com/zalando-incubator/kopf/pull/144